ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/ValueAnnotationAction.fs

```diff
@@ -0,0 +1,119 @@
+            false
+
+    let isAvailableForParameter() =
+        let parameter = dataProvider.GetSelectedElement<IParametersPatternDeclaration>()
+        isNotNull parameter &&
+        not (AnnotationUtil.isFullyAnnotatedPattern parameter.Pattern) &&
+        match parameter.Parent with
+        | :? IBinding as binding ->
+            match binding.HeadPattern with
+            | :? IFSharpDeclaration as bindingDecl ->
+                tryGetParameterAnnotationContextFromDeclaration parameter binding.ParametersDeclarationsEnumerable false bindingDecl
+            | _ ->
+                false
+
+        | :? IMemberDeclaration as memberDecl ->
+           tryGetParameterAnnotationContextFromDeclaration parameter memberDecl.ParametersDeclarationsEnumerable true memberDecl
+
+        // member val ...
+        | :? IAutoPropertyDeclaration as autoProp ->
+            // TODO: fix this one because now it seems that we can't get returnTypeInfo or namedTypeUsage from it
+            // It is not in range of declaration now
+            // Tests for member val are now disabled
+            false
+
+        | _ ->
+            false
+
+    let isAvailableForValue() =
+        let binding = dataProvider.GetSelectedElement<IBinding>()
+        let pattern = dataProvider.GetSelectedElement<IFSharpPattern>()
+
+        if isNull binding || isNull pattern then false else
+
+        let pattern =
+            match pattern.Parent with
+            | :? ITuplePat as tuplePat ->
+                tuplePat :> IFSharpPattern
+            | :? ITypedPat as typedPat ->
+                typedPat
+            | _ ->
+                pattern
+
+        let isAvailable =
+            AnnotationUtil.isValueBinding binding &&
+            StandaloneAnnotationUtil.isSupportedPatternForStandaloneAnnotation pattern
+
+        if isAvailable then
+            let forceParens =
+                pattern :? ITuplePat
+                || match binding.BindingKeyword.GetText() with
+                   | "let!" | "and!" | "use!" ->
+                       true
+                   | _ ->
+                       false
+            annotationContext <- ValueSome (ValueContext(pattern, forceParens))
+            true
+        else
+            false
+
+    override this.IsAvailable _ =
+        // this is rather complex:
+        // for parameter we can annotate anything we want basically because we can get full type of member/function
+        // for value it seems that we can annotate only ref pats because we can't really get type from wild pat, or array pat or list pat
+        // I couldn't find a way at least
+        // but check for parameter is rather long because we have to save context, or maybe it can be moved to actual psi transaction ?
+        // not sure if member declaration can return null symbolUsage
+        // if this is impossible then moving actual transition is won't be too hard
+
+        isAvailableForParameter()
+        || isAvailableForValue()
```

Comment:
This would probably be simpler as two separate actions, as it won't require neither adding a special type to distinguish the two cases, not matching it and adding a suspicious assert.